[onert] Apply nodiscard attribute

ragmani opened the pull request:
This commit applies nodiscard attribute.

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

glistening commented:
@ragmani 

1. I don't like `[[nodiscard]]` appears in almost all (even in APIs). I don't think it is the first-citizen or the most important information in our function.

2. I searched what is `[nodiscard]`. it seems a helper to give hint to compiler, maybe to get more warnings? I don't know the differences or benefits compare to as-it-is. We already get warnings if we ignores the return values with `-Wall` or static analyzer. Could you explain what is the benefit of this change?

ragmani commented:
@glistening 

Thank you for your feedback. Here are some points addressing your concerns:

1. Why `[[nodiscard]]`?
The idea behind `[[nodiscard]]` is not to make the return value “first-citizen” in every function but to protect against inadvertent mistakes. For functions that return important status codes, pointers, or computed values that could lead to resource leaks or logic errors if ignored, marking them as [[nodiscard]] enforces that the caller must explicitly consider the return value. This is particularly helpful when the function’s result isn’t automatically checked by the caller.


2. Benefits over `–Wall` or static analysis:
While it’s true that some compilers (with `-Wall`) or static analyzers may warn about unused return values, these warnings are not always issued for all user-defined functions. `[[nodiscard]]` is a standardized attribute that guarantees a warning in any compliant compiler, making the intent explicit in the code. This attribute improves documentation and communication of function contracts—anyone reading the header immediately sees that the return value should not be discarded.



In summary, the use of [[nodiscard]] improves code safety and clarity by ensuring that critical return values aren’t silently ignored. It serves as an additional layer of defense on top of existing compiler warnings and static analysis tools.

ragmani commented:
@glistening PTAL

glistening commented:
I already know what you mentioned. Still I don't think it is worth to get in. introduce `nodiscard` especially in API header (publicly exposed). I don't want to block this if 2 approvals from maintainers. (@hseok-oh and @chunseoklee)

chunseoklee reviewed (approved):
I have no objection though I do not like this look :)

ragmani commented:
> introduce `nodiscard` especially in API header (publicly exposed)

Those headers are wrapper headers, not CAPI headers. Are they exposed publicly?

glistening commented on runtime/onert/core/include/ir/train/operation/BinaryArithmetic.h:
```diff
@@ -32,7 +32,7 @@ class BinaryArithmetic : public ir::operation::BinaryArithmetic, public Trainabl
 
 public:
-  std::unique_ptr<ITrainableOperation> clone() const override;
+  [[nodiscard]] std::unique_ptr<ITrainableOperation> clone() const override;
```
Q1. `clone()` returns `unique_ptr`. Does it mean the original one is not valid after `clone()`? It is not intuitive. It may be out of the scope, but seemingly you've introduced `clone` in base class.

Q2. Is it possible not to repeat `[[nodiscard]]` by putting ``[[nodiscard]]`` only once in base class (= `ITrainableOperation`)?

ragmani replied:
> Q1. Does returning a `std::unique_ptr` from `clone()` mean the original object becomes invalid?

No, it does not. The `clone()` function creates a new copy of the object and returns a `std::unique_ptr` owning that copy. The original object remains intact and valid. The use of `std::unique_ptr` here is solely for managing the ownership of the newly created copy, not for transferring or invalidating the original object.

Q2.Is it possible not to repeat `[[nodiscard]]` by putting `[[nodiscard]]` only once in base class (= `ITrainableOperation`)?

In C++ the `[[nodiscard]]` attribute is not automatically inherited by overridden functions. Even if the virtual function in the base class is marked with `[[nodiscard]]`, each overriding function in derived classes must also be explicitly marked with `[[nodiscard]]` if you want the same warning behavior for those functions.

These points clarify that the use of `[[nodiscard]]` in each overriding implementation is required by the C++ standard and does not impact the validity of the original object returned from `clone()`.

glistening commented:
> Those headers are wrapper headers, not CAPI headers. Are they exposed publicly?

I thought you want not to allow the **API users** to ignore our API's return values.
But as you write, it is the wrapper headers.

Do you want to make sure we (internal developers) do not ignore the return values?

For most cases, each function is used only in `nnfw_api.c` like:

```c
NNFW_STATUS nnfw_prepare(nnfw_session *session)
{
  NNFW_RETURN_ERROR_IF_NULL(session);
  return session->prepare();
}

NNFW_STATUS nnfw_run(nnfw_session *session)
{
  NNFW_RETURN_ERROR_IF_NULL(session);
  return session->run();
}

NNFW_STATUS nnfw_run_async(nnfw_session *session)
{
  NNFW_RETURN_ERROR_IF_NULL(session);
  return session->run_async();
}
...
```

As I wrote at first, I don't think it is worth of the first (= the most valuable space) spaces for each changed functions signatures.

ragmani commented:
@glistening 

>I thought you want not to allow the API users to ignore our API's return values.
But as you write, it is the wrapper headers.

> Do you want to make sure we (internal developers) do not ignore the return values?

> If the latter, for most cases, each function is used only in nnfw_api.c like:

## Clarification on Header Exposure and Intent

The headers in question are indeed wrapper headers intended for internal use rather than the publicly exposed C API headers. The main intent behind adding the `[[nodiscard]]` attribute is not solely to prevent API users from ignoring return values but also to enforce rigorous error checking among **internal developers**.

## Purpose of `[[nodiscard]] `in Wrapper Headers

- **Internal Consistency and Safety**:
Even though the functions in these wrapper headers (like `nnfw_prepare`, `nnfw_run`, etc.) are typically used only within `nnfw_api.c`, marking them with `[[nodiscard]]` ensures that any mistakes in ignoring the return values are caught during development. This helps maintain robustness and reliability by making sure error conditions are properly handled.

- **Error Handling Enforcement**:
The `[[nodiscard]]` attribute acts as a safeguard. It warns developers if a function's return value, which could indicate an error, is inadvertently ignored. This is valuable not only for external API users but also for internal developers who work with the lower-level implementation details.


## Conclusion
In summary, although these headers are wrappers and not part of the public C API, applying `[[nodiscard]]` is a design decision aimed at preventing internal developers from overlooking important return values. This approach helps improve overall code safety and reliability, even if the functions are typically used in a straightforward manner within `nnfw_api.c`.

> This explanation should clarify that the use of `[[nodiscard]]` in our internal wrapper headers is intentional and valuable for maintaining high coding standards.

While it might seem that the beginning of a function signature is a “valuable space” reserved for essential keywords, the reality is that modern C++ best practices encourage using that space for meaningful annotations as well. Attributes like `[[nodiscard]]` play a crucial role in preventing programming errors by warning developers when a function’s return value is ignored, even in cases where the function is only used internally. The benefits of catching potential errors and maintaining code robustness far outweigh the minor cost of using a few extra characters at the start of the signature. In other words, the trade-off is justified because these annotations help enforce proper error handling and improve the overall quality of the code.

ragmani commented:
I personally believe that adding `[[nodiscard]]`,even if it takes up some of that “valuable” space in the function signature, provides significant benefits in terms of error prevention and code clarity. However, I understand your concerns, and if you feel that this change is not worth forcing into every function signature, I'm not inclined to push it aggressively. My intention is simply to improve our error handling practices, but I'm open to not applying this change if the trade-off doesn’t seem justified.

glistening commented:
@ragmani 

1.
It is just my personal preference.
As I wrote, 1 more approval (from maintainer) may merge this change.

2.
If it is for internal developer only, ignored return values could be checked with c++ static analyzer. I checked `cppcheck` works for detecting ignored return values.

3.
If it is for both internal developer and API user, `[[nodiscard]]` should be in public API header also.

ragmani commented:
Thank you for the feedback. Since the consensus seems to lean away from using `[[nodiscard]]`, I'll be closing this PR. I appreciate the discussion and your input on the matter.
